File: flutter_app/lib/app/grayman_app.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../core/design_system/app_colors.dart';
import '../core/networking/token_store.dart';
import '../models/worker.dart';
import 'screen.dart';
import '../features/onboarding/onboarding_screen.dart';
import '../features/auth/phone_auth_screen.dart';
import '../features/onboarding/name_entry_screen.dart';
import '../features/onboarding/avatar_picker_screen.dart';
import '../features/onboarding/choose_role_screen.dart';
import '../features/reel/record_reel_screen.dart';
import '../features/explore/reel_feed_screen.dart';
import '../features/home/home_screen.dart';
import '../features/profile/profile_screen.dart';

/// Composition root — mirrors GrayManApp.swift flat state machine.
class GrayManApp extends StatefulWidget {
  const GrayManApp({super.key});

  @override
  State<GrayManApp> createState() => _GrayManAppState();
}

class _GrayManAppState extends State<GrayManApp> {
  late AppScreen _screen;
  bool _booting = true;
  String _userName = '';
  Worker? _selectedWorker;
  AppScreen _previousScreen = AppScreen.chooseRole;
  AppScreen _reelOrigin = AppScreen.chooseRole;

  @override
  void initState() {
    super.initState();
    _screen = TokenStore.instance.token != null ? AppScreen.home : AppScreen.onboarding;
    Future.microtask(() => setState(() => _booting = false));
  }

  void _goExplore(AppScreen from) {
    setState(() {
      _previousScreen = from;
      _screen = AppScreen.explore;
    });
  }

  void _signOut() async {
    await TokenStore.instance.clear();
    setState(() {
      _userName = '';
      _selectedWorker = null;
      _previousScreen = AppScreen.chooseRole;
      _screen = AppScreen.onboarding;
    });
  }

  @override
  Widget build(BuildContext context) {
    if (_booting) {
      return const Scaffold(
        backgroundColor: AppColors.canvas,
        body: Center(child: CircularProgressIndicator(color: AppColors.shadowGrey)),
      );
    }

    return AnimatedSwitcher(
      duration: const Duration(milliseconds: 250),
      child: KeyedSubtree(key: ValueKey(_screen), child: _buildScreen()),
    );
  }

  Widget _buildScreen() {
    switch (_screen) {
      case AppScreen.onboarding:
        return OnboardingScreen(onNext: () => setState(() => _screen = AppScreen.phoneAuth));

      case AppScreen.phoneAuth:
        return PhoneAuthScreen(
          onBack: () => setState(() => _screen = AppScreen.onboarding),
          onNext: () => setState(() => _screen = AppScreen.nameEntry),
        );

      case AppScreen.nameEntry:
        return NameEntryScreen(
          onBack: () => setState(() => _screen = AppScreen.phoneAuth),
          onNext: (name) {
            setState(() {
              _userName = name;
              _screen = AppScreen.avatarOnboarding;
            });
          },
        );

      case AppScreen.avatarOnboarding:
        return AvatarPickerScreen(
          name: _userName,
          onBack: () => setState(() => _screen = AppScreen.nameEntry),
          onNext: () => setState(() => _screen = AppScreen.chooseRole),
        );

      case AppScreen.chooseRole:
        final first = _userName.split(' ').firstOrNull ?? _userName;
        return ChooseRoleScreen(
          name: first.isEmpty ? 'there' : first,
          onBack: () => setState(() => _screen = AppScreen.avatarOnboarding),
          onProfessional: () => setState(() {
            _reelOrigin = AppScreen.chooseRole;
            _screen = AppScreen.recordReel;
          }),
          onExplore: () => _goExplore(AppScreen.chooseRole),
        );

      case AppScreen.recordReel:
        return RecordReelScreen(
          onBack: () => setState(() => _screen = _reelOrigin),
          onDone: (_) => setState(() => _screen = AppScreen.home),
        );

      case AppScreen.explore:
        return ReelFeedScreen(
          onBack: () => setState(() => _screen = _previousScreen),
          onViewProfile: (w) => setState(() {
            _selectedWorker = w;
            _screen = AppScreen.workerProfile;
          }),
          onGoProfile: () => setState(() => _screen = AppScreen.profile),
          onGoHome: () => setState(() => _screen = AppScreen.home),
          onGoSettings: () => setState(() => _screen = AppScreen.profile),
        );

      case AppScreen.workerProfile:
        return ProfileScreen(
          worker: _selectedWorker,
          onBack: () => setState(() => _screen = AppScreen.explore),
        );

      case AppScreen.home:
        return HomeScreen(
          userName: _userName,
          onProfile: () => setState(() => _screen = AppScreen.profile),
          onExplore: () => _goExplore(AppScreen.home),
          onSettings: () => setState(() => _screen = AppScreen.profile),
          onViewWorker: (w) => setState(() {
            _selectedWorker = w;
            _screen = AppScreen.workerProfile;
          }),
        );

      case AppScreen.profile:
        return ProfileScreen(
          userName: _userName,
          onExplore: () => _goExplore(AppScreen.profile),
          onSignOut: _signOut,
          onRecordReel: () => setState(() {
            _reelOrigin = AppScreen.profile;
            _screen = AppScreen.recordReel;
          }),
          onHome: () => setState(() => _screen = AppScreen.home),
        );
    }
  }
}

extension _FirstOrNull<E> on List<E> {
  E? get firstOrNull => isEmpty ? null : first;
}

File: ios/GrayMan.swiftpm/App/GrayManApp.swift
import SwiftUI

// MARK: - Routing
//
// Mirrors the React App() screen-state machine. Each screen calls back
// to advance / retreat; the App owns the canonical state.

enum Screen: Equatable {
    case onboarding       // S1 — splash + Get Started / Sign In
    case phoneAuth        // S2 — phone + OTP
    case nameEntry        // SName — capture user's display name
    case avatarOnboarding // SAvatar — optional profile-photo picker
    case chooseRole       // SChoose — Professional or Explore
    case recordReel       // SRecord — 30s intro video
    case explore          // SExplore — swipeable nearby-worker feed
    case workerProfile    // S3 in "view other worker" mode
    case home             // SHome — stats card + nearby posts feed (landing)
    case profile          // S3 — own dashboard with glass tab bar
}

@main
struct GrayManApp: App {
    @State private var theme = AppTheme()
    @State private var screen: Screen
    @State private var userName: String = ""
    @State private var selectedWorker: Worker? = nil
    @State private var previousScreen: Screen = .chooseRole   // for Explore back-nav
    @State private var reelOrigin: Screen = .chooseRole       // for RecordReel back-nav (chooseRole vs profile re-shoot)

    // Composition root: swap these for real implementations as the backend lands.
    private let phoneAuthRepository: PhoneAuthRepository = APIPhoneAuthRepository()
    private let recordingService: RecordingService = AVRecordingService()

    init() {
        // Raise URLCache.shared above the iOS default (≈4 MB memory /
        // 20 MB disk) so AsyncImage's repeated fetches in the Home feed
        // hit cache after the first paint. Saves repeated 100-300 KB
        // PNG/JPEG fetches on Tier-3 networks, which is the difference
        // between "feels native" and "perpetually loading".
        URLCache.shared = URLCache(
            memoryCapacity: 32 * 1024 * 1024,    // 32 MB
            diskCapacity:   256 * 1024 * 1024,   // 256 MB
            directory: nil,
        )

        // If we already have a JWT in the Keychain, skip the auth flow and
        // drop the user straight into the Home screen. Otherwise start at
        // the splash screen.
        let hasToken = TokenStore.shared.token != nil
        _screen = State(initialValue: hasToken ? .home : .onboarding)
    }

    var body: some Scene {
        WindowGroup {
            ZStack {
                screenView
                    .transition(.opacity)

                #if DEBUG
                // Floating colour picker is a design-tool overlay; ship-mode hides it.
                ColorPanel()
                    .allowsHitTesting(true)
                #endif
            }
            .environment(theme)
            // Force light mode app-wide so `.glassEffect()` is consistent
            // regardless of the device's system appearance setting.
            .preferredColorScheme(.light)
            .dynamicTypeSize(...DynamicTypeSize.accessibility2)
            .animation(.easeInOut(duration: 0.25), value: screen)
        }
    }

    @ViewBuilder
    private var screenView: some View {
        switch screen {
        case .onboarding:
            OnboardingView(
                goNext: { screen = .phoneAuth }
            )

        case .phoneAuth:
            PhoneAuthView(
                model: PhoneAuthViewModel(repository: phoneAuthRepository),
                goBack: { screen = .onboarding },
                goNext: { screen = .nameEntry }
            )

        case .nameEntry:
            NameEntryView(
                goBack: { screen = .phoneAuth },
                goNext: { name in
                    userName = name
                    screen = .avatarOnboarding
                    Task { try? await WorkerService.shared.updateSelf(WorkerUpdateRequest(name: name)) }
                }
            )

        case .avatarOnboarding:
            AvatarPickerView(
                name: userName,
                goBack: { screen = .nameEntry },
                goNext: { screen = .chooseRole }
            )

        case .chooseRole:
            ChooseRoleView(
                name: userName.split(separator: " ").first.map(String.init) ?? userName,
                goBack: { screen = .avatarOnboarding },
                goProfessional: {
                    reelOrigin = .chooseRole
                    screen = .recordReel
                },
                goExplore: { goExplore(from: .chooseRole) }
            )

        case .recordReel:
            RecordReelView(
                model: RecordReelViewModel(service: recordingService),
                goBack: { screen = reelOrigin },
                goDone: { _ in screen = .home }
            )

        case .explore:
            ReelFeedView(
                onBack: { screen = previousScreen },
                onViewProfile: { worker in
                    selectedWorker = worker
                    screen = .workerProfile
                },
                onGoProfile: { screen = .profile },
                onGoHome: { screen = .home },
                // Settings lives inside ProfileView — route through
                // Profile so the user lands somewhere meaningful.
                onGoSettings: { screen = .profile }
            )

        case .workerProfile:
            ProfileView(
                worker: selectedWorker,
                onBack: { screen = .explore }
            )

        case .home:
            HomeView(
                userName: userName,
                onProfile: { screen = .profile },
                onExplore: { goExplore(from: .home) },
                onSettings: { screen = .profile },  // Settings lives inside ProfileView for now
                onViewWorker: { worker in
                    selectedWorker = worker
                    screen = .workerProfile
                }
            )

        case .profile:
            ProfileView(
                // Pass through whatever name we have (may be empty for a
                // returning user with a stored token). ProfileView shows
                // a redacted skeleton until the backend profile loads.
                userName: userName,
                onExplore: { goExplore(from: .profile) },
                onSignOut: {
                    userName = ""
                    selectedWorker = nil
                    previousScreen = .chooseRole
                    screen = .onboarding
                },
                onRecordReel: {
                    reelOrigin = .profile
                    screen = .recordReel
                },
                onHome: { screen = .home }
            )
        }
    }

    private func goExplore(from origin: Screen) {
        previousScreen = origin
        screen = .explore
    }
}
